Add bbc sounds provider

kieranhogg opened the pull request:
Creating a new MR for this as I seem to have failed to rebase properly and it was showing more than the expected amount of commits in the previous one (#2283).

I've addressed the original comments raised by updating the queue details to use the new method. In hindsight I should have just fixed that and merged as the rest of the functionality took a disproportionate amount of time!

## Overview
This provider now implements almost all of the full BBC Sounds functionality/catalog. The logic of whether to show the UK version or the slimmed down international version is down to the Sounds API itself. I've made this a manual toggle in the settings just until it's been a bit more road-tested.

### Existing issues and queries

- Using the new StreamMetadata works fine, the image_url doesn't seem to be used, however. I'm not sure if the display image is meant to use this yet, or not?
- I currently use some folder images from a remote URL, if there is a way to include these locally that feels a neater solution

### Future discussions
With maybe a few final tweaks, this is good enough as a starting point. Looking at another iteration, these are areas I need some guidance on whether it is worth pursuing a discussion on discord/issues to improve or adapt existing functionality, or whether they are not going to be likely or possible:

- Currently streams are set to be not seekable, ideally I'd like to look to bring pausing and seeking functionality to the stations, whether that is with the current functionality or with a custom stream implementation
- Mapping all the objects to Track, Podcast and PodcastEpisode has ended up with some odd things. So at a minimum, expanding the Radio object to include something like RadioEpisode, RadioSeries. E.g. while it is functional, browsing the API can give a list of mixed PodcastEpisodes and not being able to show which Podcast they belong to
- Linked to this, the listing in the frontend are a little limiting when trying to show more information about an item
- Logged in users are able to subscribe to radio shows and podcasts which then appear in their "My Sounds" area. At one staged I had these appear in the Library podcasts section, which worked functionally, but running the provider as a half-streaming, half-not streaming felt a bit odd. The main issue was that it surfaces the fact that radio shows are classed as podcasts so they appeared under Podcasts but I would assume users would expect them to be in Radio?

Any guidance on how welcome a discussion or plan in these areas is welcome.

kieranhogg committed:
Initial commit of BBC Sounds provider

kieranhogg committed:
Initial commit of BBC Sounds provider

kieranhogg committed:
Initial commit of BBC Sounds provider

kieranhogg committed:
Merge branch 'dev' into add-bbc-sounds-provider

kieranhogg committed:
Remove unused SUPPORTED_FEATURES

kieranhogg committed:
Remove unused SUPPORTED_FEATURES

kieranhogg committed:
Remove commented code; tidy stream and provider variables; accept ruff formatting

kieranhogg committed:
Remove commented code; tidy stream and provider variables; accept ruff formatting

kieranhogg committed:
Add search

kieranhogg committed:
Add search

kieranhogg committed:
First version of (almost) the full Sounds functionality

kieranhogg committed:
Remove accidental leading _ in get_podcast_episode()

kieranhogg committed:
Use network logo and not programme image for schedule listing

kieranhogg committed:
Remove commented out code

kieranhogg committed:
Tidy up object titles and update to latest auntie-sounds API

kieranhogg committed:
Remove unused supported features

kieranhogg committed:
Add feature flag for blank image check

kieranhogg committed:
Cancel now playing task and tidy up

kieranhogg committed:
Update icons

kieranhogg committed:
Merge branch 'add-bbc-sounds-provider' of github.com:kieranhogg/music-assistant-server into add-bbc-sounds-provider

kieranhogg committed:
Merge branch 'music-assistant:dev' into add-bbc-sounds-provider

OzGav commented:
So you weren't able to map all pre-recorded episodes to Podcast for example?

kieranhogg commented:
> So you weren't able to map all pre-recorded episodes to Podcast for example?

The vast majority of playable items have a container which does map to a Podcast or a RadioSeries (which then also maps to a Podcast eventually). There are a handful of items that are just mapped to Tracks as they don't have a parent container.

This largely works well with the content retrieved. In the below screenshot, the first row are actual podcasts, the bottom row are treated as podcats, but arguably some are more closer to a radio series, or a collection, but it still all works fine.

<img width="700" alt="image" src="https://github.com/user-attachments/assets/808c8b6d-0936-4f30-91fb-49bde35f74d5" />

At its least polished part, when you navigate to the "Latest news" menu item in the built-in API menu, you are presented with a collection of items which are mapped to Track and PodcastEpisode, but seeing them all in a jumble together highlights how much I'm bending MA's existing models and functionality to allow it to work.

<img width="700" alt="image" src="https://github.com/user-attachments/assets/c83bddb1-66ef-4d28-91cd-52be903755de" />

There's also few clunky edge cases where I can probably improve things without any extra MA changes, I just need to think of some tidier solutions. E.g. browsing radio shows from yesterday, looks mostly fine:

<img width="400" alt="image" src="https://github.com/user-attachments/assets/85549c00-d93f-4746-b46d-9a759c561e46" />

But when played the metadata is not the neatest!

<img width="600" alt="image" src="https://github.com/user-attachments/assets/2dde30bf-c9fb-403f-84d4-091184d19002" />

OzGav commented:
Yeah its tough when they don't have a coherent structure that you can work from. My Internet Archive provider would be way worse than this as it is just a bunch of people adding "stuff" in a variety of ways. At the end of the day if people can find in MA what they can find in the native app then that would seem to be a good outcome?

OzGav commented:
In regards to the logo if you can create a larger one 512x512 that would be great (wikimedia has one you could crop) and make a monochrome one as well where the three rectangles are just plain white.

OzGav commented on music_assistant/providers/bbc_sounds/manifest.json:
```diff
@@ -0,0 +1,14 @@
+  "domain": "bbc_sounds",
+  "stage": "beta",
+  "name": "BBC Sounds",
+  "description": "Listen to your favourite BBC radio stations in Music Assistant",
```
```suggestion
  "description": "Stream live BBC radio shows, podcast series and on-demand audio.",
```

kieranhogg commented:
# Docs

**FEATURES**
- Stream live BBC radio shows, podcast series and on-demand audio
- Shows the currently playing song when streaming radio stations
- Access your personalised BBC Sounds content*
- Keeps your My Sounds listen history up to date with your Music Assistant activity*


_*Requires a logged in BBC account_

**CONFIGURATION**
The provider works without a login, but you will be restricted to the limited International version of the Sounds menu

**KNOWN ISSUES/ NOTES**
The full functionality is enabled when you are a UK-based listener, and signed in to your BBC account. The BBC has slowly been moving towards excluding non-UK listeners from using its own Sounds platform, more details [here](https://www.bbc.co.uk/sounds/help/questions/listening-outside-the-uk/outside-uk-changes). This provider currently allows access to live and catch up radio to International listeners via the Sounds API, but this could be restricted in the future.

**NOT YET SUPPORTED**
- Pausing and seeking live radio
- The international menu should show avaiable podcasts, this will be supported in a future release, but is not currently
- Showing the currently playing song on catch-up
- For signed-in users, accessing your subscribed content through the library is not yet supported, but is planned

OzGav commented on music_assistant/providers/bbc_sounds/__init__.py:
```diff
@@ -0,0 +1,865 @@
+                    _Constants.CONF_RECOMMENDATIONS_HOMEPAGE,
+                ),
+                ConfigValueOption(
+                    "Show recommendations in folders in the browse page",
```
```suggestion
                    "Show recommendations in folders on the browse page",
```

OzGav commented on music_assistant/providers/bbc_sounds/__init__.py:
```diff
@@ -0,0 +1,865 @@
+        ConfigEntry(
+            key=_Constants.CONF_INTRO,
+            type=ConfigEntryType.LABEL,
+            label="A BBC Sounds account is optional, but some UK-only content may not work without",
```
```suggestion
            label="A BBC Sounds account is optional, but some UK-only content may not work without it",
```

OzGav commented on music_assistant/providers/bbc_sounds/__init__.py:
```diff
@@ -0,0 +1,865 @@
+            type=ConfigEntryType.BOOLEAN,
+            label="Enable UK Sounds content (beta)",
+            default_value=False,
+            description="Enabling this setting unlocks the full content catalog if you are a UK"
```
```suggestion
            description="Enabling this setting unlocks the full content catalog if you are a UK "
```

OzGav commented on music_assistant/providers/bbc_sounds/__init__.py:
```diff
@@ -0,0 +1,865 @@
+            label="Enable UK Sounds content (beta)",
+            default_value=False,
+            description="Enabling this setting unlocks the full content catalog if you are a UK"
+            "listener. As the API returns a wide range of media items under the same type, this "
```
Im not sure the users need to know why this is off by default? Seems like a TODO comment in the code?

kieranhogg replied:
I was just brave and removed the toggle, the provider and API should handle it without

OzGav commented on music_assistant/providers/bbc_sounds/__init__.py:
```diff
@@ -0,0 +1,865 @@
+        ConfigEntry(
+            key=_Constants.CONF_ENABLE_UK_CONTENT,
+            type=ConfigEntryType.BOOLEAN,
+            label="Enable UK Sounds content (beta)",
```
```suggestion
            label="Enable UK Sounds content",
```

The whole provider is beta :-)

OzGav commented on music_assistant/providers/bbc_sounds/__init__.py:
```diff
@@ -0,0 +1,865 @@
+    async def get_track(self, prov_track_id: str) -> Track:
+        """Get full track details by id.
+
+        Only called if provider supports ProviderFeature.LIBRARY_TRACKS.
```
You dont have this feature enabled so I don't think this function should be here?

kieranhogg replied:
Thought I was going mad and imagining it, but yes this does seem to be called even if ProviderFeature.LIBRARY_TRACKS isn't supported.

<img width="310" height="110" alt="image" src="https://github.com/user-attachments/assets/2e46096d-1edc-4a85-9ff9-404e8780cb2f" />

OzGav replied:
I'm not the expert here but I understand the LIBRARY_* features are for when the provider has a library of media items that you are syncing to MA. So if you have "tracks" in your BBC Sounds account then it makes it possible to pull those tracks across. So either it is a bug or the comment in the code is outdated...? Marcel will know!

marcelveldt replied:
that docstring is simply not right - you are probably returning a track somewhere and when those details are viewed, the core wants to fetch the full details through this method. If that is something you wont be able to handle, we'll have to look for a workaround.

kieranhogg replied:
It's the one from the demo provider so I need to delete it away. It works completely fine as is, I return Tracks which are streaming items. get_track() and get_stream_details() both work as I need them to, so unless it's problematic, no changes needed.

OzGav commented on music_assistant/providers/bbc_sounds/__init__.py:
```diff
@@ -0,0 +1,865 @@
+
+    @property
+    def supported_features(self) -> set[ProviderFeature]:
+        """Return the features supported by this Provider."""
```
Have a look at the demo provider as how the supported features is passed has changed https://github.com/music-assistant/server/blob/b67c6f882af763a7c30ace9a1de2266dd89a99c9/music_assistant/providers/_demo_music_provider/__init__.py#L100

OzGav commented on music_assistant/providers/bbc_sounds/__init__.py:
```diff
@@ -0,0 +1,865 @@
+
+    @property
+    def is_streaming_provider(self) -> bool:
+        """Return True if the provider is a streaming provider."""
```
```suggestion
        """Return True as the provider is a streaming provider."""
```

OzGav commented on music_assistant/providers/bbc_sounds/__init__.py:
```diff
@@ -0,0 +1,865 @@
+        # If we are requesting a previously-aired radio show, we lose access to the
+        # schedule time. The best we can find out from the API is original release
+        # date, so the stream title loses access to the air date
+        """Get full podcast epsisode details by id."""
```
```suggestion
        """Get full podcast episode details by id."""
```

OzGav commented on music_assistant/providers/bbc_sounds/__init__.py:
```diff
@@ -0,0 +1,865 @@
+    async def get_podcast(self, prov_podcast_id: str) -> Podcast:
+        """Get full podcast details by id.
+
+        Only called if provider supports ProviderFeature.LIBRARY_PODCASTS.
```
You should remove these comments from the model. edit: Actually if you don't have these features declared as supported then these are not required?

OzGav commented on music_assistant/providers/bbc_sounds/__init__.py:
```diff
@@ -0,0 +1,865 @@
+        is_playing: bool = False,
+    ) -> None:
+        """Handle callback when a (playable) media item has been played."""
+        # This is an OPTIONAL callback that is called when an item has been streamed.
```
You should remove these comments

OzGav commented:
@kieranhogg You can review the docs here https://github.com/music-assistant/beta.music-assistant.io/pull/49

Let me know if any changes are required!

kieranhogg committed:
Return SUPPORTED_FEATURES with instance in setup()

kieranhogg committed:
Remove the enable UK content toggle

kieranhogg committed:
Remove boiler comments and tweak wording

kieranhogg committed:
Check image_url is set after library change

kieranhogg committed:
Update library version

kieranhogg committed:
Merge branch 'dev' into add-bbc-sounds-provider

kieranhogg committed:
Merge branch 'add-bbc-sounds-provider' of github.com:kieranhogg/music-assistant-server into add-bbc-sounds-provider

kieranhogg committed:
Fix typo in comment

Co-authored-by: OzGav <gavnosp@hotmail.com>

kieranhogg commented:
> In regards to the logo if you can create a larger one 512x512 that would be great (wikimedia has one you could crop) and make a monochrome one as well where the three rectangles are just plain white.

<img width="512" height="512" alt="sounds-512" src="https://github.com/user-attachments/assets/7ba89654-eb15-4814-b673-b788c356343e" />
<img width="512" height="512" alt="sounds-mono-white" src="https://github.com/user-attachments/assets/da305e58-5849-4619-9a2e-39d5616ab952" />
<img width="512" height="512" alt="sounds-mono" src="https://github.com/user-attachments/assets/cca20fd8-1d4a-42a9-b629-d99a97c617ba" />

kieranhogg committed:
Merge branch 'dev' into add-bbc-sounds-provider

kieranhogg committed:
Remove the check to provide the seekable stream version.

It's not yet used and has the potential to affect international users, so remove it be safe.

kieranhogg committed:
Move fetching the menu to prevent it not being available if accessed by another route

kieranhogg committed:
Add an advanced option to choose a stream format

kieranhogg committed:
Update dependency version

kieranhogg commented:
I've just pushed a couple of extra tweaks for this PR. If there are no issues with using remote URLs for the images and StreamMetadata.image_url doesn't actually get used for the player image then this one is complete from my side, any further review comments aside.

marcelveldt commented:
StreamMetadata.image_url is not yet being used - the frontend needs to be adjusted still.

marcelveldt commented on music_assistant/providers/bbc_sounds/__init__.py:
```diff
@@ -0,0 +1,860 @@
+            required=True,
+        ),
+        ConfigEntry(
+            key=_Constants.CONF_STREAM_FORMAT,
+            category="advanced",
+            label="Preferred stream format",
+            type=ConfigEntryType.STRING,
+            options=[
+                ConfigValueOption(
+                    "HLS",
+                    _Constants.CONF_STREAM_FORMAT_HLS,
+                ),
+                ConfigValueOption(
+                    "MPEG-DASH",
+                    _Constants.CONF_STREAM_FORMAT_DASH,
+                ),
+            ],
+            default_value=_Constants.CONF_STREAM_FORMAT_HLS,
+            required=True,
+        ),
```
why would you make this an option and not just pick one that works best ?
If you want to keep it, I would at least set the category to "advanced"

marcelveldt commented on music_assistant/providers/bbc_sounds/__init__.py:
```diff
@@ -0,0 +1,860 @@
+            ],
+            default_value=_Constants.CONF_RECOMMENDATIONS_HOMEPAGE,
+            required=True,
+        ),
```
Why make this configurable ? Just return recommendations and if the user doesnt want them, they can hide the rows on the homescreen

marcelveldt commented on music_assistant/providers/bbc_sounds/__init__.py:
```diff
@@ -0,0 +1,860 @@
+            range=(1, 60),
+            depends_on="now_playing",
+            depends_on_value=True,
+        ),
```
why would you make this an option and not just pick a value that works best ?
If you want to keep it, I would at least set the category to "advanced"

marcelveldt commented on music_assistant/providers/bbc_sounds/__init__.py:
```diff
@@ -0,0 +1,860 @@
+                "Show details of the currently playing track instead of the station details"
+            ),
+            default_value=True,
+        ),
```
why would you make this an option ?

kieranhogg replied:
This is a holdover from when I used the incorrect method of swapping the stream details over — it appeared to be playing the song and I thought some might prefer just to see the station name. It's redundant now so I'll remove, thanks.

marcelveldt commented on music_assistant/providers/bbc_sounds/__init__.py:
```diff
@@ -0,0 +1,860 @@
+        """Return True as the provider is a streaming provider."""
+        return True
+
+    async def get_track(self, prov_track_id: str) -> Track:
```
this doesnt need any caching ?

note that you can use the @use_cache decorator

marcelveldt commented on music_assistant/providers/bbc_sounds/__init__.py:
```diff
@@ -0,0 +1,860 @@
+            raise MusicAssistantError(f"Incorrect track returned for {prov_track_id}")
+        return track
+
+    async def get_podcast_episode(self, prov_episode_id: str) -> PodcastEpisode:
```
same comment about caching

marcelveldt commented on music_assistant/providers/bbc_sounds/__init__.py:
```diff
@@ -0,0 +1,860 @@
+                        stream_details.stream_metadata.title = prev.titles["secondary"]
+                        if prev.image_url:
+                            stream_details.stream_metadata.image_url = prev.image_url
+            while True:
```
maybe you should break out this loop if "seconds_treamed" is set on the streamdetails ?
or do you have another way of aborting this task ?

kieranhogg replied:
You suggested this previously but I really struggled with using seconds_streamed to break out, although it was a while now so I can't remember exactly why. 

I cancel the tasks in on_played() which seems to work. Although as I write this out, I do wonder if this handles the scenario of multiple streams to multiple players correctly. I'll need to test.

kieranhogg replied:
I was sadly correct that my `on_played()` doesn't quite work fully. As `get_stream_details()` doesn't have any knowledge of the queue, there's no neat way of keeping track which task needs to be aborted, the current code just stores the last started stream which isn't correct.

So coming back to the original `seconds_streamed` issue in the hope of breaking out there, this is what I observe when I start, pause, then stop a stream. From your suggestion, I understand this not to be the expected behaviour?

```
2025-11-08 13:43:14.814 INFO (MainThread) [music_assistant.bbc_sounds] Seconds streamed for bbc_6music: None
2025-11-08 13:43:15.858 INFO (MainThread) [music_assistant.bbc_sounds] Seconds streamed for bbc_6music: None
2025-11-08 13:43:16.900 INFO (MainThread) [music_assistant.bbc_sounds] Seconds streamed for bbc_6music: None
2025-11-08 13:43:16.982 VERBOSE (MainThread) [music_assistant.webserver] Received: {"command":"players/cmd/power","message_id":22,"args":{"player_id":"demo_0","powered":false}}
2025-11-08 13:43:16.982 DEBUG (MainThread) [music_assistant.webserver] Handling command players/cmd/power
2025-11-08 13:43:16.985 DEBUG (MainThread) [music_assistant.players] Handling command cmd_power for player Demo Player demo_0
2025-11-08 13:43:16.986 DEBUG (MainThread) [music_assistant.players] Handling command cmd_ungroup for player Demo Player demo_0
2025-11-08 13:43:16.987 DEBUG (MainThread) [music_assistant.players] Handling command cmd_stop for player Demo Player demo_0
2025-11-08 13:43:16.988 INFO (MainThread) [music_assistant._demo_player_provider.demo_0] Received STOP command on player Demo Player demo_0
2025-11-08 13:43:17.197 INFO (MainThread) [music_assistant._demo_player_provider.demo_0] Received POWER OFF command on player Demo Player demo_0
2025-11-08 13:43:17.942 INFO (MainThread) [music_assistant.bbc_sounds] Seconds streamed for bbc_6music: None
2025-11-08 13:43:19.046 INFO (MainThread) [music_assistant.bbc_sounds] Seconds streamed for bbc_6music: None

kieranhogg replied:
I could also drop a task_id or something similar in StreamDetails.data to keep track of if that's easier than debugging the above...

marcelveldt replied:
Yeah that would be a neater way to do this, start the updater task in the get_streamdetails and then store it in the data. Then within that task, simply keep updating until seconds streamed is set. Just with a bit of debounce if playback is restarted within the first 15min.

kieranhogg replied:
See above though, seconds_streamed stays unset when I test?

MarvinSchenkel replied:
@marcelveldt Can you help out here? I think this is the last pending thing before we can merge this.

MarvinSchenkel replied:
@kieranhogg I had a chat to Marcel. He will need to add something to the core code for this. For now, we can merge your PR and add a fix for the metadata in an upcoming PR. Just checking in with you if you need to disable some code before we merge this PR while we wait for the core code to be ready?

kieranhogg replied:
The three different versions I've had for this are:

1. Disable the now playing functionality until a future update
2. Keep track of the last created task and terminate that one only, this works fine if you're streaming one station (current code)
3. Don't keep track of any tasks, they would keep running until their stream is replaced by something else (or if there is some MA garbage collection for tasks?)

Or something else temporary like keeping track of each `prov_item_id` and their task and killing them in `on_played()`, but could potentially overstep if multiple streams are using the same station?

Happy to take advice on the least bad option

MarvinSchenkel replied:
I would disable it for now and just display something static, e.g. ""BBC Sounds" or something similar. We can then fix it as soon as the core code is ready. I would like to merge this so people can start testing as we are about to start  the beta round for 2.7 ;-)

marcelveldt replied:
@kieranhogg would you be fine with a callback that gets called every X seconds in your code ?
So the idea I'm currently think of is to have an optional update_metadata callback on the streamdetails object.
You can then set that to a (async) method of your choice and it will be called periodically as long as playback is still ongoing. Then within your callback you can do sanity checks if updating is needed, set the fields on the streamdetails.stream_metadata if needed.

Would that work for you ?
we would then pass in the streamdetails to the callback to you can still track it back

kieranhogg replied:
@marcelveldt Thanks for taking some time to think it over; that feels like a neat solution. I looked at various different ways to clean up the tasks, e.g. `queue_id` from `streamdetails`, but then `on_played()` only references `prov_item_id` which isn't unique so there was always a risk when cleaning up tasks itself.

Thinking of possible edge cases, there seems to be sufficient data in `streamdetails` if in future a stream could be playing 5 minutes behind live by using `created_at + seek_position + seconds_streamed` (once I get my head around seconds_streamed...).

Non-live media metadata didn't make it into this version, but that should be even easier.

marcelveldt replied:
@kieranhogg 

I have just added stream_metadata_update_callback to the streamdetails (models bump upcoming in the server)
you can register any async (instance) method there and it will be called at each X interval with the streamdetails object at first parameter and the played seconds as second parameter

the interval can be controlled with stream_metadata_update_interval

```
StreamDetails(
    blah=blah,
    ....
    stream_metadata=StreamMetadata(initial metadata...)
    stream_metadata_update_callback = self._update_stream_metadata,
    stream_metadata_update_interval = 5  # =5 seconds
)
```

and the actual callback:
```
async def _update_stream_metadata(self, streamdetails: StreamDetails, elapsed_time: int) -> None:
    # do check if metadata needs to be updated
    # if needed, simply update the metadata fields here
    streamdetails.stream_metadata.title = blah etc
```

elapsed_time in this case will be the actual elapsed time and not the "seconds streamed" which can be a bit ahead due to buffering.

Hope this is useful this way!

@OzGav also FYI as you wanted to do something similar for the Pandora provider I believe ?

marcelveldt commented:
Converted to draft so you can work out the latest small feedback comments. Push the "ready for review" button as soon as you're ready for the final check and merge!

kieranhogg committed:
Remove unnecessary config options

kieranhogg committed:
Remove a couple of incorrectly-set StreamDetails variables on Radio streams

kieranhogg committed:
Tidy up after removing config options

kieranhogg committed:
Add some sensible caching

MarvinSchenkel commented on music_assistant/providers/bbc_sounds/__init__.py:
```diff
@@ -0,0 +1,800 @@
+
+        return results
+
+    async def get_podcast(self, prov_podcast_id: str) -> Podcast:
```
We can also use caching here

MarvinSchenkel commented:
In general, I think we can use some more caching and longer cache periods. Have a look at the other providers, e.g. [Qobuz](https://github.com/music-assistant/server/blob/dev/music_assistant/providers/qobuz/__init__.py) and [YTM](https://github.com/music-assistant/server/blob/dev/music_assistant/providers/ytmusic/__init__.py). Other than that, I think this is good to be merged.

kieranhogg committed:
Increase cache expiry

kieranhogg committed:
Add cache to get_podcast()

kieranhogg committed:
Disable 'now playing' until supported in core

MarvinSchenkel reviewed (approved):
Thanks @kieranhogg for your hard work 🎉 !

OzGav commented:
@kieranhogg Just checking the provider against the docs and I note there are a couple of changes to the settings. One question, what is the "Show local radio stations" toggle for?
